pkg/compute/logstream/completed_streamer.go

```diff
@@ -0,0 +1,61 @@
+package logstream
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/bacalhau-project/bacalhau/pkg/lib/concurrency"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+)
+
+type CompletedStreamerParams struct {
+	Execution *models.Execution
+}
+
+// CompletedStreamer is a streamer for completed executions that streams the
+// output from the execution's RunOutput field to the channel.
+type CompletedStreamer struct {
+	execution *models.Execution
+}
+
+func NewCompletedStreamer(params CompletedStreamerParams) *CompletedStreamer {
+	return &CompletedStreamer{
+		execution: params.Execution,
+	}
+}
+
+func (s *CompletedStreamer) Stream(ctx context.Context) chan *concurrency.AsyncResult[models.ExecutionLog] {
+	ch := make(chan *concurrency.AsyncResult[models.ExecutionLog])
+	go func() {
+		defer close(ch)
+		if s.execution.RunOutput != nil {
+			s.process(ch, s.execution.RunOutput.STDOUT, models.ExecutionLogTypeSTDOUT)
+			s.process(ch, s.execution.RunOutput.STDERR, models.ExecutionLogTypeSTDERR)
+		}
+	}()
+	return ch
+}
+
+func (s *CompletedStreamer) process(ch chan *concurrency.AsyncResult[models.ExecutionLog], output string, typ models.ExecutionLogType) {
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	for scanner.Scan() {
+		asyncResult := concurrency.AsyncResult[models.ExecutionLog]{
+			Value: models.ExecutionLog{
+				Type: typ,
+				Line: scanner.Text() + "\n",
+			},
+		}
+		ch <- &asyncResult
+	}
+	if scanner.Err() != nil {
+		asyncResult := concurrency.AsyncResult[models.ExecutionLog]{
+			Err: fmt.Errorf("failed to read output: %w", scanner.Err()),
+		}
+		ch <- &asyncResult
+	}
+}
```

Comment:
Good point. That would allow `process` to exit if the channel was full and the context is done, right?

Reply:
Done in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/80d1eef8857a74bd9ceb967d9958041bbcd42308

Reply:
> That would allow process to exit if the channel was full and the context is done, right?

If the channel is full it will block (thus my comments about buffers) if ctx is done it exits.